rfd/0078-role-variables.md

```diff
@@ -0,0 +1,264 @@
+claims coming from internal and external user traits.
+
+Requirements:
+
+1. Variable definitions should support some form of pattern matching so that
+   traits can be easily and powerfully extended.
+2. A variable defined in one role should be usable in many other roles.
+
+Non-requirements:
+
+1. Variables can depend on other variables.
+
+## Why
+
+Variables will be useful to specify additional traits that are not available in
+OIDC/SAML claims.
+
+Teleport admins don't always have control over their org's OIDC/SAML provider.
+
+Role variables will simplify role evaluation, keep logic in one place, and
+minimize repetition.
+
+## Details
+
+Example role spec:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: node_users
+spec:
+  vars:
+    # Each variable has a name it will be referenced by.
+    - name: "logins"
+      # values hold a list of possible values which will all be appended to
+      # form the complete list of values for the variable.
+      values:
+        # The simplest value is a list of static strings.
+        - out: [ubuntu]
+
+        # Values can also contain traits and transforms.
+        - out: ['{{regexp.replace(external.username, "-", "_")}}']
+
+        # Values can also take an input, which will usually be a trait.
+        - input: "{{external.email}}"
+          # match is a regular expression which must match the input in order
+          # for this value to be included.
+          match: "^nic@goteleport.com$"
+          out: [root]
+
+        - input: "{{external.email}}"
+          # match may contain contain capture groups which can be expanded in
+          # the output.
+          #
+          # Here "alice@goteleport.com" maps to "alice", but "foo@example.com"
+          # will be rejected.
+          match: "^(.*)@goteleport.com$"
+          out: ["$1"]
+
+    - name: "allow-env"
+      values:
+        # An input will expand to a list of N inputs if the trait is a list of size N.
+        # Any input value which matches the `match` regex will contribute one
+        # output value, which may contain regex captures from that specific input.
+        #
+        # Here [devs, env-staging, env-prod] maps to [staging, prod].
+        - input: "{{external.groups}}"
+          match: '^env-(\w+)$'
+          out: ['$1']
+
+        # If any of the N input values match the `not_match` regex, none of the
+        # values will be used.
+        #
+        # Here [devs] maps to [dev], but [devs, contractors] maps to []
+        - input: "{{external.groups}}"
+          match: '^devs$'
+          not_match: 'contractors'
+          out: ["dev"]
+  allow:
+    logins:
+      # Variables can be referenced as `vars.<role_name>.<variable_name>` or
+      # `vars["<role_name>"]["<variable_name"]`.
+      - "{{vars.node_users.logins}}"
+    node_labels:
+      env: ["{{vars.node_users["allow-env"]}}"]
+```
+
+These variables could be used in another role without redefinition:
```

Comment:
Let's keep variables role-local to avoid confusing behavior. Today, each role is self-sufficient and it's update is atomic. Role-local variables will keep the same guarantee. The downside is that some roles will have to declare the same variables many times. But I think it's OK for now.